File: scripts/testRetrieval.js
import dotenv from 'dotenv';
import path from 'path';
import { fileURLToPath } from 'url';
import { createEmbedding } from '../lib/embeddings.js';
import { queryVectors, getIndexStats } from '../lib/pinecone.js';

// Load environment variables
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
dotenv.config({ path: path.join(__dirname, '..', '.env.local') });

/**
 * Display search results in a formatted way
 */
function displayResults(results, query) {
  console.log('\n' + '='.repeat(80));
  console.log(`🔍 Query: "${query}"`);
  console.log('='.repeat(80) + '\n');

  if (results.length === 0) {
    console.log('❌ No results found\n');
    return;
  }

  results.forEach((result, index) => {
    const metadata = result.metadata || {};
    const score = (result.score * 100).toFixed(2);
    
    console.log(`📌 Result ${index + 1} (Similarity: ${score}%)`);
    console.log('─'.repeat(80));
    
    if (metadata.type === 'code') {
      console.log(`📄 Type: Code`);
      console.log(`📁 File: ${metadata.filename || 'Unknown'}`);
      console.log(`🔗 URL: ${metadata.url || 'N/A'}`);
      console.log(`📊 Size: ${metadata.size || 'N/A'} bytes`);
    } else if (metadata.type === 'documentation') {
      console.log(`📄 Type: Documentation`);
      console.log(`📋 Title: ${metadata.title || 'Unknown'}`);
      console.log(`📁 File: ${metadata.filename || 'Unknown'}`);
    }
    
    console.log(`🕐 Indexed: ${metadata.indexedAt || 'N/A'}`);
    
    if (metadata.content) {
      const preview = metadata.content.length > 300 
        ? metadata.content.substring(0, 300) + '...' 
        : metadata.content;
      console.log(`\n📝 Content Preview:\n${preview}`);
    }
    
    console.log('\n');
  });
}

/**
 * Test retrieval with a query
 */
async function testQuery(query, topK = 5, filterType = null) {
  try {
    // Create embedding for the query
    console.log('🔄 Creating query embedding...');
    const queryEmbedding = await createEmbedding(query);
    
    // Query Pinecone
    console.log('🔍 Searching Pinecone...');
    const filter = filterType ? { type: filterType } : {};
    const results = await queryVectors(queryEmbedding, topK, filter);
    
    // Display results
    displayResults(results, query);
    
    return results;
  } catch (error) {
    console.error('❌ Error during query:', error.message);
    throw error;
  }
}

/**
 * Run multiple test queries
 */
async function runTestSuite() {
  console.log('🧪 Starting RAG System Test Suite\n');
  
  // Get index stats first
  console.log('📊 Fetching index statistics...\n');
  const stats = await getIndexStats();
  
  console.log('='.repeat(80));
  console.log('INDEX STATISTICS');
  console.log('='.repeat(80));
  console.log(`💾 Total Vectors: ${stats.totalRecordCount || 0}`);
  console.log(`📦 Dimension: ${stats.dimension || 'N/A'}`);
  console.log('='.repeat(80));

  if (!stats.totalRecordCount || stats.totalRecordCount === 0) {
    console.log('\n⚠️  Index is empty! Please run indexing scripts first:');
    console.log('   npm run index:codebase');
    console.log('   npm run index:docs\n');
    return;
  }

  // Test queries
  const testQueries = [
    {
      name: 'General Code Search',
      query: 'React component that handles user authentication',
      topK: 3,
    },
    {
      name: 'Documentation Search',
      query: 'business logic and requirements',
      topK: 3,
      filter: 'documentation',
    },
    {
      name: 'Code Pattern Search',
      query: 'API calls and error handling',
      topK: 3,
      filter: 'code',
    },
    {
      name: 'Specific Feature',
      query: 'state management and hooks',
      topK: 3,
    },
  ];

  console.log('\n🚀 Running Test Queries...\n');

  for (const test of testQueries) {
    console.log(`\n${'▶'.repeat(40)}`);
    console.log(`TEST: ${test.name}`);
    console.log('▶'.repeat(40));
    
    await testQuery(test.query, test.topK, test.filter);
    
    // Small delay between queries
    await new Promise(resolve => setTimeout(resolve, 1000));
  }

  console.log('\n' + '='.repeat(80));
  console.log('✅ TEST SUITE COMPLETE');
  console.log('='.repeat(80) + '\n');
}

/**
 * Interactive mode - query from command line
 */
async function interactiveMode() {
  const query = process.argv.slice(2).join(' ');
  
  if (!query) {
    console.log('❌ Please provide a query');
    console.log('Usage: npm run test:retrieval -- your search query here\n');
    return;
  }

  console.log('🔍 Interactive Query Mode\n');
  await testQuery(query, 5);
}

/**
 * Main function
 */
async function main() {
  const args = process.argv.slice(2);
  
  if (args.length > 0) {
    // Interactive mode with query
    await interactiveMode();
  } else {
    // Run full test suite
    await runTestSuite();
  }
}

// Run the script
main().catch(error => {
  console.error('\n❌ Fatal Error:', error.message);
  process.exit(1);
});

File: python/scripts/test_retrieval.py
#!/usr/bin/env python3
"""
Test RAG retrieval system.
Run predefined test queries or interactive queries against Pinecone.
"""

import logging
import sys
import time
from pathlib import Path
from typing import List, Dict, Any, Optional

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from dotenv import load_dotenv

from src.core.embeddings import create_embedding
from src.core.pinecone_client import query_vectors, get_index_stats

# Load environment variables
load_dotenv()

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
)
logger = logging.getLogger(__name__)

# Content preview length
PREVIEW_LENGTH = 300


def display_results(results: List[Dict[str, Any]], query: str) -> None:
    """
    Display search results in a formatted way.

    Args:
        results: List of search results from Pinecone.
        query: The search query.
    """
    print("\n" + "=" * 80)
    print(f'Query: "{query}"')
    print("=" * 80 + "\n")

    if not results:
        print("No results found\n")
        return

    for index, result in enumerate(results):
        metadata = result.get("metadata", {})
        score = result.get("score", 0)
        score_pct = f"{score * 100:.2f}"

        print(f"Result {index + 1} (Similarity: {score_pct}%)")
        print("-" * 80)

        result_type = metadata.get("type", "unknown")

        if result_type == "code":
            print(f"Type: Code")
            print(f"File: {metadata.get('filename', 'Unknown')}")
            print(f"URL: {metadata.get('url', 'N/A')}")
            print(f"Size: {metadata.get('size', 'N/A')} bytes")
        elif result_type == "documentation":
            print(f"Type: Documentation")
            print(f"Title: {metadata.get('title', 'Unknown')}")
            print(f"File: {metadata.get('filename', 'Unknown')}")
        else:
            # Local codebase format
            print(f"Type: {metadata.get('fileType', 'Unknown')}")
            print(f"File: {metadata.get('filePath', 'Unknown')}")
            if metadata.get("startLine"):
                print(f"Lines: {metadata.get('startLine')}-{metadata.get('endLine')}")

        print(f"Indexed: {metadata.get('indexedAt', 'N/A')}")

        content = metadata.get("content", "")
        if content:
            preview = content[:PREVIEW_LENGTH] + "..." if len(content) > PREVIEW_LENGTH else content
            print(f"\nContent Preview:\n{preview}")

        print("\n")


def test_query(
    query: str,
    top_k: int = 5,
    filter_type: Optional[str] = None,
) -> List[Dict[str, Any]]:
    """
    Test retrieval with a query.

    Args:
        query: Search query.
        top_k: Number of results to return.
        filter_type: Optional filter for type (code/documentation).

    Returns:
        List of search results.
    """
    try:
        # Create embedding for the query
        print("Creating query embedding...")
        query_embedding = create_embedding(query)

        # Query Pinecone
        print("Searching Pinecone...")
        filter_dict = {"type": filter_type} if filter_type else None
        results = query_vectors(query_embedding, top_k=top_k, filter=filter_dict)

        # Display results
        display_results(results, query)

        return results
    except Exception as e:
        print(f"Error during query: {e}")
        raise


def run_test_suite() -> None:
    """Run multiple test queries."""
    print("Starting RAG System Test Suite\n")

    # Get index stats first
    print("Fetching index statistics...\n")
    stats = get_index_stats()

    print("=" * 80)
    print("INDEX STATISTICS")
    print("=" * 80)
    print(f"Total Vectors: {stats.get('total_vector_count', 0)}")
    print(f"Dimension: {stats.get('dimension', 'N/A')}")
    print("=" * 80)

    total_count = stats.get("total_vector_count", 0)
    if not total_count:
        print("\nIndex is empty! Please run indexing scripts first:")
        print("   python scripts/index_codebase.py")
        print("   python scripts/index_documentation.py\n")
        return

    # Test queries
    test_queries = [
        {
            "name": "General Code Search",
            "query": "React component that handles user authentication",
            "top_k": 3,
            "filter": None,
        },
        {
            "name": "Documentation Search",
            "query": "business logic and requirements",
            "top_k": 3,
            "filter": "documentation",
        },
        {
            "name": "Code Pattern Search",
            "query": "API calls and error handling",
            "top_k": 3,
            "filter": "code",
        },
        {
            "name": "Specific Feature",
            "query": "state management and hooks",
            "top_k": 3,
            "filter": None,
        },
    ]

    print("\nRunning Test Queries...\n")

    for test in test_queries:
        print(f"\n{'>' * 40}")
        print(f"TEST: {test['name']}")
        print(">" * 40)

        test_query(test["query"], test["top_k"], test["filter"])

        # Small delay between queries
        time.sleep(1)

    print("\n" + "=" * 80)
    print("TEST SUITE COMPLETE")
    print("=" * 80 + "\n")


def interactive_mode() -> None:
    """Interactive mode - query from command line."""
    query = " ".join(sys.argv[1:])

    if not query:
        print("Please provide a query")
        print("Usage: python test_retrieval.py <your search query here>\n")
        return

    print("Interactive Query Mode\n")
    test_query(query, top_k=5)


def main():
    """Main function."""
    args = sys.argv[1:]

    if args:
        # Interactive mode with query
        interactive_mode()
    else:
        # Run full test suite
        run_test_suite()


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        logger.error(f"Fatal Error: {e}")
        sys.exit(1)
